README.md
# Lenovo-OpenAI-vscExtension

An extension that uses OpenAI's API in vscode. 2023.4.4 I started to develop.

一个在vscode中使用OpenAI的API的扩展插件。 2023.4.4由本人在北京联想开始着手开发。

##!!注意，目前v2版本需要挂梯子!!

## Features

###V2

* ctrl+shift+p输入Lenovo即可出现多条功能命令
* 右键菜单也有多条功能命令
* 扩展webview视图交互UI（设置APIkey，清空面板，询问OpenAI）

![v2](https://lenovo-openai-demo.oss-cn-beijing.aliyuncs.com/v2.gif)

备注：使用功能命令之前，如果没有选取代码块将询问整个文件

###V1

安装此插件后，可以在文件中用鼠标选取一部分文本，然后ctrl+shift+p输入openAI: chat命令，即可等待chatGPT对你的回复。

![demo-gif](https://lenovo-openai-demo.oss-cn-beijing.aliyuncs.com/demo.gif)

## Requirements

vscode + install this extension


## Extension Settings

**不需要OpenAI账户APIkey，Enjoy**!

(最新的v2需要APIKey)


## Release Notes

## V2 可视化交互UI，右键菜单，用户自设置APIKey

### 2.0.1

2023.4.17 询问方式全部转成中文，以便结果也是中文

### 2.0.0

2023.4.17 new version, 可视化交互UI，右键菜单，用户自设置APIKey


## V1 代理+微服务，0门槛上ChatGPT

### 1.1.0

2023.4.14 updated, 右键菜单，新文件打开

### 1.0.1

2023.4.10 patched，内部使用

### 1.0.0

2023.4.7 Initial，chat功能


CHANGELOG.md
# Change Log


## V2 可视化交互UI，右键菜单，用户自设置APIKey

### 2.0.0

2023.4.17 new version, 可视化交互UI，右键菜单，用户自设置APIKey



## V1 代理+微服务，0门槛上ChatGPT

### 1.1.0

2023.4.14 updated, 右键菜单，新文件打开

### 1.0.1

2023.4.10 patched，内部使用

### 1.0.0

2023.4.7 Initial，chat功能


test/test.js
/***
 * @creater:ACBash
 * @create_time:23-4-4 15:12:48
 * @last_modify:ACBash
 * @modify_time:23-4-19 16:5:39
 * @line_count:11
 **/

//目前比较流行的前端构建工具有哪些？
//目前比较流行的低代码平台运用到了哪些技术？
//麻烦说说湖北武汉每个区的平均房价
/* let a = 1;
console.log(a);
console.log(exports);
console.log(module.exports);
console.log(module); */
console.log(exports.__esModule);
exports.ax = 1;
// 新建一个

openAI.js
/***
 * @creater:ACBash
 * @create_time:23-4-4 15:27:49
 * @last_modify:ACBash
 * @modify_time:23-4-8 14:9:7
 * @line_count:124
 **/

/* const axios = require('axios');     //网络请求调用openAI的API
const openai = require('openai');   //引入openAI的SDK

// var apiKey = 'YOUR_API_KEY_HERE';    openAI_APIkey

// let chatGPT = null;

const configuration = new openai.Configuration({ apiKey: apiKey });
const client = new openai.OpenAIApi(configuration);

const result = async (content) => {
    return await client.createChatCompletion({
        // 使用当前 OpenAI 开放的最新 4 模型，如果后续 新模型 发布，则修改此处参数即可
        model: 'gpt-4',
        messages: [{ role: 'user', content }]
    });
};

console.log(result("hello!")); */

/* chatGPT = async (content) => {
    return await client.createChatCompletion({
        // 使用当前 OpenAI 开放的最新 4 模型，如果后续 新模型 发布，则修改此处参数即可
        model: 'gpt-4',
        messages: [{ role: 'user', content }]
    });
}; */

//用户输入
// let userInput = "hello!";

//将用户具体消息发送给 ChatGPT
// const result = chatGPT(userInput);

//打印ChatGPT的反馈
// console.log(result);

/* function callApi(query, callback) {
    var options = {
        url: 'https://api.openai.com/v1/chat/completions',
        headers: {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer ' + apiKey
        },
        method: 'POST',
        body: JSON.stringify({
            "model": "gpt-4",
            "message": [{"role": "user", "content": query}]
        }),
        json: {
            'prompt': query,
            'max_tokens': 100,
            'temperature': 0.5,
            'top_p': 1,
            'frequency_penalty': 0,
            'presence_penalty': 0
        }
    };
    request(options, function(error, response, body) {
        if (!error && response.statusCode === 200) {
            callback(body);
        } else {
            console.log('Error contacting OpenAI: ' + error);
        }
    });
}

callApi("hello!", console.log); */

/* module.exports = {
    callApi: callApi
}; */

/* const { Configuration, OpenAIApi } = require("openai");

const configuration = new Configuration({
  apiKey: "",
});
const openai = new OpenAIApi(configuration);

const start = async function(){
    const completion = await openai.createChatCompletion({
        model: "gpt-3.5-turbo",
        messages: [{role: "user", content: "Hello world"}],
    });
    console.log(completion.data.choices[0].message);
};

start(); */

const axios = require('axios');
const host = "微服务接口地址";
const service = axios.create({
    baseURL: host,
    timeout: '60000'
});

const request = (options) => {
    //以post方式发起get请求
    if(options.method == 'get'){
        options.params = options.data;
    }
    return service(options);
};
//vue框架和react有什么区别？
const chat = async (data) => {
    const res = await request({
        method: 'get',
        url: '/hello',
        data
    });
    /* console.log(res); */
    return res;
};

let userInput = {
    "content": "hello!",
};

/* chat(userInput); */

module.exports = {
    chat: chat
};

aircode.js
/***
 * @creater:ACBash
 * @create_time:23-4-4 17:25:7
 * @last_modify:ACBash
 * @modify_time:23-4-8 14:10:22
 * @line_count:34
 **/

// @see https://docs.aircode.io/guide/functions/
const aircode = require('aircode');

const { Configuration, OpenAIApi } = require("openai");

const configuration = new Configuration({
  apiKey: "your_openAI_APIkey",
});
const openai = new OpenAIApi(configuration);

/* const start = async function(){
    const completion = await openai.createChatCompletion({
        model: "gpt-3.5-turbo",
        messages: [{role: "user", content: "Hello world"}],
    });
    console.log(completion.data.choices[0].message);
};

start(); */

module.exports = async function(params, context) {
  console.log('Received params:', params);

  const completion = await openai.createChatCompletion({
        model: "gpt-3.5-turbo",
        messages: [{role: "user", content: "Hello world"}],
    });
  console.log(completion.data.choices[0].message);
  
  return {
    message: completion.data.choices[0].message,
  };
}


media/main.js
/***
 * @creater:ACBash
 * @create_time:23-4-17 10:26:59
 * @last_modify:ACBash
 * @modify_time:23-4-17 15:8:26
 * @line_count:76
 **/

// This script will be run within the webview itself
// It cannot access the main VS Code APIs directly.
(function () {
    const vscode = acquireVsCodeApi();
    const list = document.getElementById("qa-list");

    // Handle messages sent from the extension to the webview
    window.addEventListener("message", (event) => {
        const message = event.data;

        switch (message.type) {
            case "addQuestion":
                const html = message.code != null
                    ? marked.parseInline(message.value + "<br /> <br /><pre class='overflow-auto'><code>```" + message.code + "```</code></pre>")
                    : message.value;

                list.innerHTML +=
                    `<div class="p-4 self-end mb-4">
                        <p class="font-bold mb-5 flex">
                            <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="w-5 h-5 mr-2"><path stroke-linecap="round" stroke-linejoin="round" d="M17.982 18.725A7.488 7.488 0 0012 15.75a7.488 7.488 0 00-5.982 2.975m11.963 0a9 9 0 10-11.963 0m11.963 0A8.966 8.966 0 0112 21a8.966 8.966 0 01-5.982-2.275M15 9.75a3 3 0 11-6 0 3 3 0 016 0z" /></svg>
                            You
                        </p>
                        <div>${html}</div>
                    </div>`;

                document.getElementById("in-progress")?.classList?.remove("hidden");
                list.scrollTo(0, list.scrollHeight);
                break;
            case "addResponse":
                document.getElementById("in-progress")?.classList?.add("hidden");

                list.innerHTML +=
                    `<div class="p-4 self-end mb-4 pb-8">
                        <p class="font-bold mb-5 flex">
                            <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="w-5 h-5 mr-2"><path stroke-linecap="round" stroke-linejoin="round" d="M8.25 3v1.5M4.5 8.25H3m18 0h-1.5M4.5 12H3m18 0h-1.5m-15 3.75H3m18 0h-1.5M8.25 19.5V21M12 3v1.5m0 15V21m3.75-18v1.5m0 15V21m-9-1.5h10.5a2.25 2.25 0 002.25-2.25V6.75a2.25 2.25 0 00-2.25-2.25H6.75A2.25 2.25 0 004.5 6.75v10.5a2.25 2.25 0 002.25 2.25zm.75-12h9v9h-9v-9z" /></svg>
                            LenovoGPT
                        </p>
                        <div>${marked.parse(message.value)}</div>
                    </div>`;

                list.scrollTo(0, list.scrollHeight);
                break;
            case "clearHTML":
                list.innerHTML = "";
                break;
            default:
                break;
        }
    });

    let submitHandler = function (e) {
        e.preventDefault();
        e.stopPropagation();
        const input = document.getElementById("question-input");
        if (input.value?.length > 0) {
            vscode.postMessage({
                type: "askChatGPT",
                value: input.value,
            });

            input.value = "";
        }
    };

    document.getElementById("clear-button")?.addEventListener("click", () => {
        list.innerHTML = "";
        vscode.postMessage({ type: "clearChat", });
    });
    document.getElementById("ask-button")?.addEventListener("click", submitHandler);
    document.getElementById("question-input")?.addEventListener("keydown", function (e) {
        console.log(e.key);
        if (e.key === "Enter" && !e.shiftKey) {
            submitHandler(e);
        }
    });
})();

Lenovo-OpenAI.js
/***
 * @creater:ACBash
 * @create_time:23-4-10 14:37:43
 * @last_modify:ACBash
 * @modify_time:23-4-19 15:25:24
 * @line_count:77
 **/

//引入vscode扩展API
const vscode = require('vscode');

//引入webview-provider
const { ChatGptViewProvider } = require("./chatgpt-view-provider");

//vscode扩展上下文的激活/未激活函数
function activate(context) {
    //显示已激活扩展
    console.log(`Good! 联想内部使用OpenAI插件成功激活!`);
    vscode.window.showInformationMessage(`Good! 联想内部使用OpenAI插件成功激活!`);
    
    //侧边栏可视化交互式窗口UI
    const chatViewProvider = new ChatGptViewProvider(context);
    
    context.subscriptions.push(
		vscode.commands.registerCommand('Lenovo-OpenAI.askGPT', askChatGPT),
		vscode.commands.registerCommand('Lenovo-OpenAI.whyBroken', askGPTWhyBroken),
		vscode.commands.registerCommand('Lenovo-OpenAI.explainPls', askGPTToExplain),
		vscode.commands.registerCommand('Lenovo-OpenAI.refactor', askGPTToRefactor),
		vscode.commands.registerCommand('Lenovo-OpenAI.clear', clearHTML),
		vscode.commands.registerCommand('Lenovo-OpenAI.resetToken', resetToken),
		vscode.commands.registerCommand('Lenovo-OpenAI.resetAPIKey', resetAPIKey),
		vscode.window.registerWebviewViewProvider("Lenovo-OpenAI.view", chatViewProvider, {
			webviewOptions: { retainContextWhenHidden: true }
		})
	);

    async function resetToken() {
		await context.globalState.update('chatgpt-session-token', null);
		await context.globalState.update('chatgpt-clearance-token', null);
		await context.globalState.update('chatgpt-user-agent', null);
		await chatViewProvider.ensureApiKey();
		// await vscode.window.showInformationMessage("Token reset, you'll be prompted for it next to you next ask ChatGPT a question.");
	}

    async function resetAPIKey() {
        await chatViewProvider.resetApiKey();
    }

    async function askChatGPT(userInput) {
        if (!userInput) {
            userInput = await vscode.window.showInputBox({ prompt: "向ChatGPT问一个问题" });
        }

        let editor = vscode.window.activeTextEditor;

        if (editor) {
            const selectedCode = editor.document.getText(vscode.window.activeTextEditor?.selection);
            const entireFileContents = editor.document.getText();

            const code = selectedCode
                ? selectedCode
                : `这是${editor.document.languageId}文件并且其中的内容是\n\n${entireFileContents}`;

            chatViewProvider.sendOpenAiApiRequest(userInput, code);
        }
    }

    async function askGPTToExplain() { await askChatGPT('你能解释一下这段代码是做什么的吗?'); }
	async function askGPTWhyBroken() { await askChatGPT('为什么这段代码会报错?'); }
	async function askGPTToRefactor() { await askChatGPT('你能重构优化这段代码吗'); }

    async function clearHTML() { await chatViewProvider.clearHTML(); }
}

function deactivate(context) {
    //显示未激活扩展
    console.log(`GodDamnIt! 插件未激活`);
    vscode.window.showInformationMessage(`GodDamnIt! 插件未激活`);
}

//模块导出
module.exports = {
    activate,
    deactivate
};

v1.0.1/extension.js
/***
 * @creater:ACBash
 * @create_time:23-4-4 16:20:29
 * @last_modify:ACBash
 * @modify_time:23-4-7 12:14:13
 * @line_count:51
 **/

// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require('vscode');
const openAI = require('./openAI.js');

// 从编辑器中获取选取的文本
const getSelectedText = () => {
	return vscode.window.activeTextEditor?.document.getText(vscode.window.activeTextEditor?.selection) || '';
};

// This method is called when your extension is activated
// Your extension is activated the very first time the command is executed

/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "demo" is now active!');

	let chatCommand = vscode.commands.registerCommand('demo.chat', async () => {
		let selectedText = getSelectedText();
		if(selectedText == ''){
			vscode.window.showErrorMessage('请先选取文本!');
			return;
		}

		let userQuery = {"content": `${selectedText}`};

		const res = await openAI.chat(userQuery);
		const ans = res.data.message.content;
		console.log(ans);

		vscode.window.showInformationMessage(`${ans}`);
	});
	context.subscriptions.push(chatCommand);

	let helloCommand = vscode.commands.registerCommand('demo.helloWorld', function () {
		vscode.window.showInformationMessage('Hello openAI from demo!');
	});
	context.subscriptions.push(helloCommand);
}

// This method is called when your extension is deactivated
function deactivate() {}

module.exports = {
	activate,
	deactivate
}


chatgpt-view-provider.js
/***
 * @creater:ACBash
 * @create_time:23-4-12 16:36:5
 * @last_modify:ACBash
 * @modify_time:23-4-17 15:19:51
 * @line_count:195
 **/

//webview-HTML
/* `<!DOCTYPE html>
<html>
<head>
    <title>固定在页面底部的输入框</title>
    <style type="text/css">
        body {
            margin: 0;
            padding: 0;
            height: 100vh;
            display: flex;
            flex-direction: row;
            justify-content: center;
            align-items: center;
        }
        .container {
            position: fixed;
            bottom: 0;
            width: 100%;
            padding: 10px;
            box-sizing: border-box;
            background-color: #1E1E1E;
            text-align: center;
        }
        .container input[type=text] {
            background-color: #3C3C3C;
            color: #fff;
            padding: 10px;
            width: 70%;
            border: none;
            border-radius: 5px;
            outline: none;
            font-size: 16px;
            box-sizing: border-box;
            vertical-align: middle;
        }
        .container input[type=submit] {
            padding: 10px 20px;
            background-color: #4CAF50;
            border: none;
            border-radius: 5px;
            color: #fff;
            font-size: 16px;
            cursor: pointer;
            vertical-align: middle;
        }
    </style>
</head>
<body>
    <div class = 'chat'>
        1111111111111
    </div>
    <div class="container">
        <form action="#" method="post">
            <input type="text" name="input" placeholder="请输入内容...">
            <input type="submit" value="提交">
        </form>
    </div>
</body>
</html>`; */

const vscode = require('vscode');
const request = require('./request');

class ChatGptViewProvider {
    constructor(context) {
        this.webView = null;
        this.context = context;
        this.message = null;
        this.apiKey = null;
    }

    resolveWebviewView(webviewView, _context, _token) {
        this.webView = webviewView;

        webviewView.webview.options = {
            enableScripts: true,
            localResourceRoots: [this.context.extensionUri],
        };

        webviewView.webview.html = this.getHtml(webviewView.webview);

        webviewView.webview.onDidReceiveMessage(data => {
            if (data.type === 'askChatGPT') {
                this.sendOpenAiApiRequest(data.value);
            }
        });

        if (this.message !== null) {
            this.sendMessageToWebView(this.message);
            this.message = null;
        }
    }

    getHtml(webview) {
        const scriptUri = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'media', 'main.js'));
        const stylesMainUri = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'media', 'main.css'));
        
        return `<!DOCTYPE html>
                <html lang="en">
                <head>
                    <meta charset="UTF-8">
                    <meta name="viewport" content="width=device-width, initial-scale=1.0">
                    <link href="${stylesMainUri}" rel="stylesheet">
                    <script src="https://cdn.jsdelivr.net/npm/marked/marked.min.js"></script>
                    <script src="https://cdn.tailwindcss.com"></script>
                </head>
                <body class="overflow-hidden">
                    <div class="flex flex-col h-screen">
                        <div class="flex-1 overflow-y-auto" id="qa-list"></div>
                        <div id="in-progress" class="p-4 flex items-center hidden">
                            <div style="text-align: center;">
                                <div>Please wait while we handle your request ❤️</div>
                                <div class="loader"></div>
                            </div>
                        </div>
                        <div class="p-4 flex items-center">
                            <div class="flex-1">
                                <textarea
                                    type="text"
                                    rows="2"
                                    class="border p-2 w-full"
                                    id="question-input"
                                    placeholder="请输入问题..."
                                ></textarea>
                            </div>
                            <button style="background: var(--vscode-button-background)" id="ask-button" class="p-2 ml-5">提交</button>
                        </div>
                    </div>
                    <script src="${scriptUri}"></script>
                </body>
                </html>`;
    }

    sendMessageToWebView(message) {
        if (this.webView) {
            this.webView?.webview.postMessage(message);
        } else {
            this.message = message;
        }
    }

    async resetApiKey() {
        const apiKeyInput = await vscode.window.showInputBox({
            title: "设定你的OpenAI-APIKey",
            prompt: "请输入你的OpenAI-APIKey. 你可以在https://platform.openai.com/account/api-keys找到",
            ignoreFocusOut: true,
            placeHolder: "sk-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
        });
        this.apiKey = apiKeyInput;
        this.context.globalState.update('chatgpt-api-key', this.apiKey);
    }

    async ensureApiKey() {
        this.apiKey = await this.context.globalState.get('chatgpt-api-key');

        if (!this.apiKey) {
            const apiKeyInput = await vscode.window.showInputBox({
                title: "设定你的OpenAI-APIKey",
                prompt: "请输入你的OpenAI-APIKey. 你可以在https://platform.openai.com/account/api-keys找到",
                ignoreFocusOut: true,
                placeHolder: "sk-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
            });
            this.apiKey = apiKeyInput;
            this.context.globalState.update('chatgpt-api-key', this.apiKey);
        }
    }

    async sendOpenAiApiRequest(prompt, code) {
        await this.ensureApiKey();

        const question = (code) ? `${prompt}: ${code}` : prompt;

        if (!this.webView) {
            await vscode.commands.executeCommand('chatgpt-vscode-plugin.view.focus');
        } else {
            this.webView?.show?.(true);
        }

        this.sendMessageToWebView({ type: 'addQuestion', value: prompt, code });

        let response = await request(this.apiKey, question);
        let ans = response.message.content;

        this.sendMessageToWebView( { type: 'addResponse', value: ans } );
    }

    async clearHTML(){
        this.webView.webview.postMessage({ type: 'clearHTML' });
    }
}

module.exports = {
    ChatGptViewProvider
};